### PR TITLE
[SPARK-23817][SQL]Migrate ORC file format read path to data source V2

### PR DESCRIPTION
**Pull request by gengliangwang**
## What changes were proposed in this pull request?
Migrate ORC file format read path to data source V2. 

### Supports:

1. Scan ColumnarBatch
2. Scan UnsafeRow
3. Push down filters
4. Push down required columns

### Not supported( due to limitation of data source V2):

1. Read multiple file path
2. Read bucketed file.

## How was this patch tested?

unit test

### CODE REVIEW

File: sql/core/src/main/java/org/apache/spark/sql/execution/datasources/orc/OrcColumnarBatchReader.java [hunk 1]
```diff
@@ -62,6 +62,12 @@
    */
   private int[] requestedColIds;
 
+  /**
+   * The column IDs of the ORC file partition schema which are required by this reader.
+   * -1 means this required column doesn't exist in the ORC partition columns.
+   */
+  private int[] requestedPartitionColds;
```
Comment (gatorsmile) [inline sql/core/src/main/java/org/apache/spark/sql/execution/datasources/orc/OrcColumnarBatchReader.java]: requestedPartitionColds -> requestedPartitionColIds

**Comment by gatorsmile** [comment]
Let us do some tests by changing `spark.sql.sources.default` to `orc` and see whether all the tests can pass.

**Comment by gengliangwang** [comment]
@jose-torres @cloud-fan

File: sql/core/src/main/scala/org/apache/spark/sql/DataFrameReader.scala [hunk 2]
```diff
@@ -187,6 +189,14 @@ class DataFrameReader private[sql](sparkSession: SparkSession) extends Logging {
...
     }
 
+    // SPARK-23817 Since datasource V2 didn't support reading multiple files yet,
+    // ORC V2 is only used when loading single file path.
+    val allPaths = CaseInsensitiveMap(extraOptions.toMap).get("path") ++ paths
+    val orcV2 = OrcDataSourceV2.satisfy(sparkSession, source, allPaths.toSeq)
+    if (orcV2.isDefined) {
+      option("path", allPaths.head)
+      source = orcV2.get
+    }
```
Comment (jose-torres) [inline sql/core/src/main/scala/org/apache/spark/sql/DataFrameReader.scala]: It seems weird that DataFrameReader is modified here. Will DataSourceV2 implementations generally need to modify DataFrameReader, or is it just a temporary hack because of the mentioned lack of support? In the latter case, is there a plan to add this support soon?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/DataSourceScanExec.scala [hunk 3]
```diff
@@ -368,8 +368,7 @@ case class FileSourceScanExec(
     val bucketed =
       selectedPartitions.flatMap { p =>
         p.files.map { f =>
-          val hosts = getBlockHosts(getBlockLocations(f), 0, f.getLen)
```
Comment (jose-torres) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/DataSourceScanExec.scala]: Would it make sense to split the refactoring changes into their own PR? It's hard to tell at a glance which parts of the change are refactoring and which are new V2 implementation.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceStrategy.scala [hunk 4]
```diff
@@ -213,6 +215,26 @@ case class DataSourceAnalysis(conf: SQLConf) extends Rule[LogicalPlan] with Cast
   }
 }
 
+/**
+ * Replaces [[OrcDataSourceV2]] with [[DataSource]] if parent node is [[InsertIntoTable]].
+ * This is because [[OrcDataSourceV2]] doesn't support writing data yet.
```
Comment (jose-torres) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceStrategy.scala]: Shouldn't Spark handle this already when it sees that OrcDataSourceV2 doesn't implement WriteSupport?

**Review by jose-torres — COMMENTED** [review]
A few high level questions.

File: sql/core/src/main/scala/org/apache/spark/sql/DataFrameReader.scala (hunk 2 above)
Comment (gengliangwang) [reply sql/core/src/main/scala/org/apache/spark/sql/DataFrameReader.scala]: This is temporary hack. I think @cloud-fan will create a PR to support reading multiple files recently.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/DataSourceScanExec.scala (hunk 3 above)
Comment (gengliangwang) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/DataSourceScanExec.scala]: Yes that is a good idea.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceStrategy.scala (hunk 4 above)
Comment (gengliangwang) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceStrategy.scala]: Not for `InsertIntoTable`.

Comment (jose-torres) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceStrategy.scala]: Is this also a temporary hack, then? It seems like Spark should know it can't write to a source which doesn't implement WriteSupport, no matter what the shape of the query performing the write is.

File: sql/core/src/main/java/org/apache/spark/sql/execution/datasources/orc/OrcColumnarBatchReader.java [hunk 5]
```diff
@@ -146,22 +152,19 @@ public void initialize(
    */
   public void initBatch(
       TypeDescription orcSchema,
-      int[] requestedColIds,
       StructField[] requiredFields,
-      StructType partitionSchema,
+      int[] requestedColIds,
+      int[] requestedPartitionColIds,
```
Comment (cloud-fan) [inline sql/core/src/main/java/org/apache/spark/sql/execution/datasources/orc/OrcColumnarBatchReader.java]: param doc for them.

File: sql/core/src/main/java/org/apache/spark/sql/execution/datasources/orc/OrcColumnarBatchReader.java [hunk 6]
```diff
@@ -171,47 +174,39 @@ public void initBatch(
 
       // Initialize the missing columns once.
```
Comment (cloud-fan) [inline sql/core/src/main/java/org/apache/spark/sql/execution/datasources/orc/OrcColumnarBatchReader.java]: it's not only missing columns not, but also partition columns

File: sql/core/src/main/scala/org/apache/spark/sql/DataFrameReader.scala (hunk 2 above)
Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/DataFrameReader.scala]: yea it's a temporary hack. We will support multi-path soon.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/DataSourceScanExec.scala (hunk 3 above)
Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/DataSourceScanExec.scala]: I'm afraid these refactors only make sense in this PR, for reusing code between v1 and v2.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceStrategy.scala (hunk 4 above)
Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceStrategy.scala]: This is a tempory hack. v1 has the same problem: when inserting into a table which is backed by a non-writable data source, Spark would fail during planning.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/orc/OrcUnsafeRowReaderFactory.scala [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+import org.apache.spark.sql.catalyst.InternalRow
+import org.apache.spark.sql.catalyst.expressions.{JoinedRow, UnsafeRow}
+import org.apache.spark.sql.catalyst.expressions.codegen.GenerateUnsafeProjection
+import org.apache.spark.sql.execution.datasources._
+import org.apache.spark.sql.execution.datasources.orc.{OrcDeserializer, OrcUtils}
+import org.apache.spark.sql.sources.v2.reader.{DataReader, DataReaderFactory}
+import org.apache.spark.sql.types.StructType
+import org.apache.spark.util.SerializableConfiguration
+
+case class OrcUnsafeRowReaderFactory(
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/orc/OrcUnsafeRowReaderFactory.scala]: there are a lot of duplication between this and `OrcBatchDataReaderFactory`, and we may have more when migrating other file formats.

File: sql/core/src/main/scala/org/apache/spark/sql/DataFrameReader.scala (hunk 2 above)
Comment (jose-torres) [reply sql/core/src/main/scala/org/apache/spark/sql/DataFrameReader.scala]: What about bucketed reads? WIll they need a similar change here, or is that lack of support handled elsewhere? (Or am I misunderstanding something about that part of the description - I'm not super familiar with the ORC source)

**Comment by cloud-fan** [comment]
also cc @rdblue

**Comment by gengliangwang** [comment]
retest this please.

**Comment by gengliangwang** [comment]
retest this please.

**Comment by cloud-fan** [comment]
retest this please

File: sql/core/src/main/scala/org/apache/spark/sql/DataFrameReader.scala (hunk 2 above)
Comment (rdblue) [reply sql/core/src/main/scala/org/apache/spark/sql/DataFrameReader.scala]: How does v2 not support reading multiple files?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/DataSourceScanExec.scala (hunk 3 above)
Comment (rdblue) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/DataSourceScanExec.scala]: Why does this only make sense for this PR? It looks like this is a reasonable refactor that could be stand-alone.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/DataSourceScanExec.scala [hunk 8]
```diff
@@ -396,107 +395,34 @@ case class FileSourceScanExec(
...
-    val defaultMaxSplitBytes =
-      fsRelation.sparkSession.sessionState.conf.filesMaxPartitionBytes
+    val maxSplitBytes = PartitionedFileUtil.maxSplitBytes(relation.sparkSession, selectedPartitions)
     val openCostInBytes = fsRelation.sparkSession.sessionState.conf.filesOpenCostInBytes
-    val defaultParallelism = fsRelation.sparkSession.sparkContext.defaultParallelism
-    val totalBytes = selectedPartitions.flatMap(_.files.map(_.getLen + openCostInBytes)).sum
-    val bytesPerCore = totalBytes / defaultParallelism
-
-    val maxSplitBytes = Math.min(defaultMaxSplitBytes, Math.max(openCostInBytes, bytesPerCore))
     logInfo(s"Planning scan with bin packing, max size: $maxSplitBytes bytes, " +
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/DataSourceScanExec.scala]: We can also move this log to `PartitionedFileUtil.maxSplitBytes`

File: sql/core/src/main/scala/org/apache/spark/sql/execution/DataSourceScanExec.scala [hunk 9]
```diff
@@ -396,107 +395,34 @@ case class FileSourceScanExec(
...
-            val hosts = getBlockHosts(blockLocations, offset, size)
-            PartitionedFile(
-              partition.values, file.getPath.toUri.toString, offset, size, hosts)
-          }
-        } else {
-          val hosts = getBlockHosts(blockLocations, 0, file.getLen)
-          Seq(PartitionedFile(
-            partition.values, file.getPath.toUri.toString, 0, file.getLen, hosts))
-        }
+        // getPath() is very expensive so we only want to call it once in this block:
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/DataSourceScanExec.scala]: this can be taken care of in `PartitionedFileUtil.splitFiles`

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceStrategy.scala [hunk 10]
```diff
@@ -213,6 +215,26 @@ case class DataSourceAnalysis(conf: SQLConf) extends Rule[LogicalPlan] with Cast
...
+ * @param sparkSession
+ */
+class FallBackToOrcV1(sparkSession: SparkSession) extends Rule[LogicalPlan] {
+  private def convertToOrcV1(v2Relation: DataSourceV2Relation): LogicalPlan = {
+    val v1 = DataSource.apply(
+      sparkSession = sparkSession,
+      paths = Seq.empty,
+      userSpecifiedSchema = v2Relation.userSpecifiedSchema,
+      className = "org.apache.spark.sql.execution.datasources.orc",
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceStrategy.scala]: don't hardcode it

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FilePartitionUtil.scala [hunk 11]
```diff
@@ -0,0 +1,225 @@
...
+import org.apache.spark.TaskContext
+import org.apache.spark.deploy.SparkHadoopUtil
+import org.apache.spark.internal.Logging
+import org.apache.spark.rdd.InputFileBlockHolder
+import org.apache.spark.sql.SparkSession
+import org.apache.spark.sql.catalyst.InternalRow
+import org.apache.spark.sql.vectorized.ColumnarBatch
+import org.apache.spark.util.NextIterator
+
+object FilePartitionUtil extends Logging {
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FilePartitionUtil.scala]: why we have both `PartitionedFileUtil` and `FilePartitionUtil`?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/FileReaderFactory.scala [hunk 12]
```diff
@@ -0,0 +1,50 @@
...
+ */
+
+package org.apache.spark.sql.execution.datasources.v2
+
+import org.apache.spark.TaskContext
+import org.apache.spark.sql.catalyst.InternalRow
+import org.apache.spark.sql.execution.datasources.{FilePartition, FilePartitionUtil, PartitionedFile}
+import org.apache.spark.sql.sources.v2.reader.{DataReader, DataReaderFactory}
+
+case class FileReaderFactory[T](
```
Comment (rdblue) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/FileReaderFactory.scala]: Why is this class public? Isn't this internal to HadoopFsRelation's v2 implementation?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/FileReaderFactory.scala [hunk 13]
```diff
@@ -0,0 +1,50 @@
...
+  override def preferredLocations(): Array[String] = {
+    FilePartitionUtil.getPreferredLocations(file)
+  }
+}
+
+case class InternalRowDataReader[T](iter: Iterator[InternalRow])
+  extends DataReader[T] {
+  override def next(): Boolean = iter.hasNext
+
+  override def get(): T = iter.next().asInstanceOf[T]
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/FileReaderFactory.scala]: explain why this works, i.e. we use type erase hack to return columnar batch.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/orc/OrcDataSourceV2.scala [hunk 14]
```diff
@@ -0,0 +1,194 @@
...
+  }
+
+  override def createReader(schema: StructType, options: DataSourceOptions): DataSourceReader = {
+    new OrcDataSourceReader(options, Some(schema))
+  }
+}
+
+case class OrcDataSourceReader(options: DataSourceOptions, userSpecifiedSchema: Option[StructType])
+  extends ColumnarBatchFileSourceReader
+  with SupportsPushDownCatalystFilters {
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/orc/OrcDataSourceV2.scala]: which file source doesn't support this? I think all file sources support partitioning.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/orc/OrcPartitionDiscoverySuite.scala [hunk 15]
```diff
@@ -66,13 +68,6 @@ abstract class OrcPartitionDiscoveryTest extends OrcTest {
       spark.read.orc(base.getCanonicalPath).createOrReplaceTempView("t")
 
       withTempTable("t") {
-        checkAnswer(
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/orc/OrcPartitionDiscoverySuite.scala]: we should not remove tests

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/orc/OrcDataSourceV2.scala [hunk 16]
```diff
@@ -0,0 +1,194 @@
...
+
+  override def inferSchema(files: Seq[FileStatus]): Option[StructType] = {
+    OrcUtils.readSchema(sparkSession, files)
+  }
+
+  private var pushedFiltersArray: Array[Expression] = Array.empty
+
+  override def readFunction: PartitionedFile => Iterator[InternalRow] = {
```
Comment (rdblue) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/orc/OrcDataSourceV2.scala]: I think it is a bad idea to continue using `PartitionedFile => Iterator[InternalRow]` in v2.

I understand not wanting to change much about how this works, just to get the code behind the v2 API. But this pattern is broken and causes resource problems that the v2 API nudges implementers to fix.

What resource problems? This doesn't implement close properly, forcing close to happen at task end by calling functions registered when files are opened. We've gone back through and replaced the iterators with closeable versions so that we release resources more quickly because the callback-based close does not scale.

I would like to see this problem fixed instead of copying it into v2.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala [hunk 17]
```diff
@@ -1185,6 +1185,13 @@ object SQLConf {
       .stringConf
       .createWithDefault("")
 
+  val DISABLED_V2_DATA_SOURCE_READERS = buildConf("spark.sql.disabledV2DataSourceReaders")
```
[the remark this answers is not in the document]
Comment (gengliangwang) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala]: This is following `disabledV2StreamingMicroBatchReaders`.  And currently this PR only supports reading.

File: sql/core/src/main/java/org/apache/spark/sql/execution/datasources/orc/OrcColumnarBatchReader.java [hunk 18]
```diff
@@ -233,7 +234,7 @@ private boolean nextBatch() throws IOException {
 
     if (!copyToSpark) {
       for (int i = 0; i < requiredFields.length; i++) {
-        if (requestedColIds[i] != -1) {
+        if (requestedColIds[i] != -1 && requestedPartitionColIds[i] == -1) {
```
[the remark this answers is not in the document]
Comment (gengliangwang) [reply sql/core/src/main/java/org/apache/spark/sql/execution/datasources/orc/OrcColumnarBatchReader.java]: Yes

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FilePartitionUtil.scala (hunk 11 above)
Comment (gengliangwang) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FilePartitionUtil.scala]: `PartitionedFileUtil` is about how we get `PartitionedFile`.
`FilePartitionUtil` is about how we get `FilePartition` and convert them to `InternalRow`

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/orc/OrcDataSourceV2.scala (hunk 14 above)
Comment (gengliangwang) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/orc/OrcDataSourceV2.scala]: I thought some of the file source would choose `SupportsPushDownFilters` instead of `SupportsPushDownCatalystFilters`. Not very sure about this.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/orc/OrcDataSourceV2.scala (hunk 16 above)
Comment (gengliangwang) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/orc/OrcDataSourceV2.scala]: Yes, I am quite frustrated when I update the code and use `PartitionedFile => Iterator[InternalRow]` as V1 did. 
I was trying to reduce duplicated code between vectorized reader and unsafe row reader. And we can reuse the code in `FileScanRDD`.
I know this makes the V2 implementation meaningless. I will keep finding a good solution.

File: sql/core/src/main/java/org/apache/spark/sql/execution/datasources/orc/OrcColumnarBatchReader.java (hunk 18 above)
[the remark this answers is not in the document]
Comment (cloud-fan) [reply sql/core/src/main/java/org/apache/spark/sql/execution/datasources/orc/OrcColumnarBatchReader.java]: we should also apply this check in the `copyToSpark` branch

[the remark this answers is not in the document]
Comment (cloud-fan) [reply sql/core/src/main/java/org/apache/spark/sql/execution/datasources/orc/OrcColumnarBatchReader.java]: also add comment for this.

File: sql/core/src/main/scala/org/apache/spark/sql/DataFrameReader.scala (hunk 2 above)
Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/DataFrameReader.scala]: because we need to define how to pass multiple paths via options. I have a PR to fix it, I'll bring it update to date.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/DataSourceScanExec.scala (hunk 3 above)
Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/DataSourceScanExec.scala]: moving code need a reason. The reason here is to help us to reuse the code. But if we do it in another PR, what is the reason? It doesn't make the code more clear IMO.

Comment (rdblue) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/DataSourceScanExec.scala]: Better organization to support other changes like this one is the reason.

@jose-torres was right to point out that these changes are self-contained enough to go in a separate PR and @gengliangwang and I both agreed. Why make this commit larger than necessary?

**Comment by gengliangwang** [comment]
Discussed with @cloud-fan offline. The conclusion is that we can use simple factory pattern for data factory, so that we can avoid redundant code easily and stop using `PartitionedFile => Iterator[InternalRow]`.
He will make a PR soon. I will continue updating this one after his PR merged.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/DataSourceScanExec.scala (hunk 3 above)
Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/DataSourceScanExec.scala]: if we agree that a separated PR is self-contained as it can help this PR, I'm also OK with it.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala (hunk 17 above)
[the remark this answers is not in the document]
Comment (gatorsmile) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala]: We need a better name.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/DataSourceScanExec.scala (hunk 3 above)
Comment (rdblue) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/DataSourceScanExec.scala]: Yeah, I think the commit itself would be self-contained reorganization. The motivation is to refactor for this PR, which is okay.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/orc/OrcDataSourceV2.scala (hunk 16 above)
Comment (gengliangwang) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/orc/OrcDataSourceV2.scala]: With #21029, we can get rid of this.

**Comment by gengliangwang** [comment]
retest this please.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FilePartitionUtil.scala [hunk 19]
```diff
@@ -0,0 +1,225 @@
...
+      }
+      // Add the given file to the current partition.
+      currentSize += file.length + openCostInBytes
+      currentFiles += file
+    }
+    closePartition()
+    partitions
+  }
+
+  def compute(
```
[the remark this answers is not in the document]
Comment (tengpeng) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FilePartitionUtil.scala]: Reasonable concern. Is it because the author followed the convention in RDD in which `compute` is used?

File: sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala [hunk 20]
```diff
@@ -241,39 +240,47 @@ final class DataFrameWriter[T] private[sql](ds: Dataset[T]) {
...
+        val options = new DataSourceOptions((extraOptions ++
+          DataSourceV2Utils.extractSessionConfigs(
+            ds = ds.asInstanceOf[DataSourceV2],
+            conf = df.sparkSession.sessionState.conf)).asJava)
+        // Using a timestamp and a random UUID to distinguish different writing jobs. This is good
+        // enough as there won't be tons of writing jobs created at the same second.
+        val jobId = new SimpleDateFormat("yyyyMMddHHmmss", Locale.US)
+          .format(new Date()) + "-" + UUID.randomUUID()
+        val writer = ds.asInstanceOf[WriteSupport]
+          .createWriter(jobId, df.logicalPlan.schema, mode, options)
```
Comment (tengpeng) [inline sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala]: I am not sure I understand this: why do use `.createWriter` here, but we do not use `.createReader` in `DataFrameReader`. It seems "unsymmetrical" to me.

**Comment by gengliangwang** [comment]
Status update: we are working on new proposal for changing the Data source API, to resolve the problems exposed in this PR.
Before the new proposal is adopted or denied, this PR remains open.
